libs/autotagging-commons/src/main/java/org/opensearch/rule/InMemoryRuleProcessingService.java

```diff
@@ -0,0 +1,147 @@
+     *                           matching rule
+     * @return a label if there is unique label otherwise empty
+     */
+    public Optional<String> evaluateLabel(List<AttributeExtractor<String>> attributeExtractors) {
+        assert attributeValueStoreFactory != null;
+        Optional<String> result = Optional.empty();
+        for (AttributeExtractor<String> attributeExtractor : attributeExtractors) {
+            AttributeValueStore<String, String> valueStore = attributeValueStoreFactory.getAttributeValueStore(
+                attributeExtractor.getAttribute()
+            );
+            for (String value : attributeExtractor.extract()) {
+                Optional<String> possibleMatch = valueStore.get(value);
+
+                if (possibleMatch.isEmpty()) {
+                    return Optional.empty();
+                }
+
+                if (result.isEmpty()) {
+                    result = possibleMatch;
+                } else {
+                    boolean isThePossibleMatchEqualResult = possibleMatch.get().equals(result.get());
+                    if (!isThePossibleMatchEqualResult) {
+                        return Optional.empty();
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    private static abstract class RuleProcessingOperation {
+        protected final Rule rule;
+        protected final AttributeValueStoreFactory attributeValueStoreFactory;
+
+        public RuleProcessingOperation(Rule rule, AttributeValueStoreFactory attributeValueStoreFactory) {
+            this.rule = rule;
+            this.attributeValueStoreFactory = attributeValueStoreFactory;
+        }
+
+        void perform() {
+            final FeatureType feature = rule.getFeatureType();
+            final String label = rule.getFeatureValue();
+
+            for (Map.Entry<Attribute, Set<String>> attributeEntry : rule.getAttributeMap().entrySet()) {
+                processAttributeEntry(attributeEntry);
+            }
+        }
+
+        protected AttributeValueStore<String, String> getAttributeValueStore(final Attribute attribute) {
+            return this.attributeValueStoreFactory.getAttributeValueStore(attribute);
+        }
+
+        protected abstract void processAttributeEntry(Map.Entry<Attribute, Set<String>> attributeEntry);
+    }
+
+    private static class DeleteRuleOperation extends RuleProcessingOperation {
+        public DeleteRuleOperation(Rule rule, AttributeValueStoreFactory attributeValueStoreFactory) {
+            super(rule, attributeValueStoreFactory);
+        }
+
+        @Override
+        protected void processAttributeEntry(Map.Entry<Attribute, Set<String>> attributeEntry) {
+            AttributeValueStore<String, String> valueStore = getAttributeValueStore(attributeEntry.getKey());
+            for (String value : attributeEntry.getValue()) {
+                valueStore.remove(value);
+            }
```

Comment:
What if there are no more attributeValues for particular attribute? Should we remove the attributeStore itself from the map?

Reply:
I don't think removing the `AttributeValueStore` is correct in my opinion since we initialize them at the bootup time only and the instances are limited and very less in number.